sdks/python/apache_beam/pipeline.py

```diff
@@ -919,11 +918,49 @@ def visit_transform(self, transform_node):
         requirements=context.requirements())
     proto.components.transforms[root_transform_id].unique_name = (
         root_transform_id)
+    self.merge_compatible_environments(proto)
     if return_context:
       return proto, context  # type: ignore  # too complicated for now
     else:
       return proto
 
+  @staticmethod
+  def merge_compatible_environments(proto):
+    """Tries to minimize the number of distinct environments by merging
+    those that are compatible (currently defined as identical).
+
+    Mutates proto as contexts may have references to proto.components.
+    """
+    env_map = {}
+    canonical_env = {}
+    files_by_hash = {}
+    for env_id, env in proto.components.environments.items():
+      # First deduplicate any file dependencies by their hash.
+      for dep in env.dependencies:
+        if dep.type_urn == common_urns.artifact_types.FILE.urn:
+          file_payload = beam_runner_api_pb2.ArtifactFilePayload.FromString(
+              dep.type_payload)
+          if file_payload.sha256:
+            if file_payload.sha256 in files_by_hash:
+              file_payload.path = files_by_hash[file_payload.sha256]
```

Comment:
FYI, we de-duplicate artifacts again before staging: https://github.com/apache/beam/blob/ba62704a4640979763655a25fd1d603e60bc8e80/sdks/python/apache_beam/runners/dataflow/internal/apiclient.py#L663

Reply:
Yeah. We could probably clean that up a lot once Runner v1 goes away.